### ISSUE TITLE
TypeError when input simulation enabled

### ISSUE THREAD

**Issue by vi7Quuce** [dessant/buster#139]
**System**

* OS name: Fedora
* OS version: 31
* Browser name: Firefox 
* Browser version: 72.0.2

**Extension**

* Extension version: 0.7.0

* User input simulation: yes

* Client app version: 0.3.0
* Client app installed successfully: yes

**Bug description**

When user input simulation is enabled, Buster does not work. I get an error in the console (see Logs below). As far as I can tell, the latest client app was installed correctly. When I disable user input simulation, Buster works as expected with my own API key.

**Logs**

Browser:
```
TypeError: o is undefined			 script.js:1:83802
```

Setup log:
```
2020/02/09 14:06:56.653511 Starting setup (version: 0.3.0)
2020/02/09 14:06:57.304205 Loading setup page
2020/02/09 14:06:57.763902 Getting install location
2020/02/09 14:07:00.822209 Installing client
2020/02/09 14:07:01.119558 Closing setup
```

Client app:
```
2020/02/09 14:10:20.728975 Starting client (version: 0.3.0)
2020/02/09 14:10:20.728994 Receiving message
2020/02/09 14:10:20.729079 Processing message
2020/02/09 14:10:20.729083 Command: ping
2020/02/09 14:10:20.729085 Sending response
2020/02/09 14:10:20.729119 Receiving message
2020/02/09 14:10:20.767716 Processing message
2020/02/09 14:10:20.767727 Command: moveMouse
2020/02/09 14:10:20.767731 X: 140, Y: 577
2020/02/09 14:10:20.770235 Sending response
2020/02/09 14:10:20.770283 Receiving message
2020/02/09 14:10:20.878110 Processing message
2020/02/09 14:10:20.878125 Command: clickMouse
2020/02/09 14:10:20.888750 Sending response
2020/02/09 14:10:20.888815 Receiving message
2020/02/09 14:10:30.898014 Closing client

```

**Comment by dessant** [comment]
Thanks for the bug report! Which API are you using? Do you also get an error when you use the client app with the `Wit Speech API (managed)` service?

**Comment by tao-oat** [comment]
Hi @dessant ! When I added this issue I was using the Google Cloud Speech API, but I get the same error with the managed Wit Speech API.

**Comment by dessant** [comment]
I haven't been able to reproduce the issue so far. Do you have any other extensions installed that might interfere? It would be useful to test the extension with a brand new browser profile.

I'll release the extension with source maps going forward to have more usable error messages.

**Comment by dessant** [comment]
I've released 0.7.1 with source maps, please update before testing. 

https://addons.mozilla.org/en-US/firefox/addon/buster-captcha-solver/

Before triggering the error, visit about:debugging#/runtime/this-firefox and search for Buster, then click on the Inspect button. Viewing the console there might give you a more usable traceback, and correct line numbers for the exception. The line numbers are also clickable, so you'll be able to see where the error was thrown.

**Comment by dessant** [comment]
Please also describe what are you seeing during the test. Does the mouse move to the correct button, and is it clicked? Does the browser show a notification about an error, or is the error visible in the console only and the extension just hangs?

**Comment by tao-oat** [comment]
Thanks for responding so quickly to this bug report! I'm now using Buster 0.7.1 in a completely new profile, but with the same settings (i.e. Google Cloud Speech API & user input simulation enabled). For what it's worth, I'm testing using https://patrickhlauke.github.io/recaptcha/.

Even with no other extensions, I get the same error. First, here's what I see: I click the recaptcha checkbox, the visual challenge pops up, and I click the Buster icon. It shows the loading indicator for 10 seconds before returning to the Buster icon and showing a desktop notification (from Buster) saying `Something went wrong. Open the browser console for details.`. The browser console just shows the same `TypeError` as earlier.

The console from `about:debugging` doesn't update at all when this error occurs. It shows the following lines (from before testing the recaptcha):
```
Could not map contract ID '@mozilla.org/sharepicker;1' to CID {1201d357-8417-4926-a694-e6408fbedcf8} because no implementation of the CID is registered.
Migrating storage (sync): undefined => DlgF14Chrh index.js:28:10
Applying revision (sync): UoT3kGyBH - Initial version index.js:32:12
Applying revision (sync): ONiJBs00o - Add IBM Watson Speech to Text API index.js:32:12
Applying revision (sync): UidMDYaYA - Add Microsoft Azure Speech to Text API index.js:32:12
Applying revision (sync): nOedd0Txqd - Add Wit Speech API and tryEnglishSpeechModel option index.js:32:12
Applying revision (sync): ZtLMLoh1ag - Add loadEnglishChallenge option index.js:32:12
Applying revision (sync): t335iRDhZ8 - Add simulateUserInput option index.js:32:12
Applying revision (sync): X3djS8vZC - Add autoUpdateClientApp option index.js:32:12
Applying revision (sync): DlgF14Chrh - Revision description
```
And then nothing else.

Two thoughts:
* How does the client simulator backend interact with the browser extension? I'm wondering if some firewall rules may be the issue.
* In this new version, the error occurs at position `script.js:1:83787`. Since the error doesn't seem to be appearing source-mapped in the extension console, perhaps there's a manual way to convert this to a code location?

**Comment by dessant** [comment]
It sounds like the mouse is not moving at all and the extension cannot find the expected audio source, so it throws an error. I've just remembered reading some time ago that Fedora uses Wayland by default, and it looks like robotgo does not yet support Wayland. :disappointed:

You could try disabling Wayland, if that's acceptable for your use case: https://github.com/go-vgo/robotgo/issues/178

The issue mentions that keyboard events are working as expected, you could also try navigating to the extension button with Tab, and pressing Enter. If solving is initiated with a keyboard, the extension will use keyboard events to navigate instead of the mouse. Though it's possible that there's a bigger chance of being temporarily blocked form accessing the audio challenge when keyboard navigation is used.

Thanks a bunch for continuing to debug the issue, your input has been valuable! I'll add a note to the installation guide about Wayland.

**Comment by tao-oat** [comment]
Ah, that explains it! Thanks a lot for your help :) I was running into quite a few issues trying to record my screen on Wayland to demonstrate what the error looked like, so maybe I should've guessed...

I'll keep using Buster without user input simulation or with the keyboard workaround you mentioned. Feel free to close this issue!

**Comment by adamwinn** [comment]
I'm seeing this too on Mac 10.15.3, Firefox 74.0b5, Buster 0.7.1
`TypeError: o is undefined`

**Comment by dsedivec** [comment]
Also seeing this on Mac.  macOS 10.14.6, Firefox 74.0b7, Buster 0.7.1.  No mouse movement, just times out.  I get the "TypeError: o is undefined" in the browser console, nothing when I "inspect" the Buster add-on.

**Comment by dessant** [comment]
@adamwinn, @dsedivec this issue is about Wayland support on Linux.

On macOS the lack of mouse movement could be caused by the security measures of the operating system, please follow the installation guide and let me know if it helped.  The client app is released untested for macOS because I don't have an Apple device, so your feedback is valuable and could contribute to a better installation guide.

> Go to System Preferences > Security & Privacy > Privacy > Accessibility, and allow the browser to control your computer in order to enable the client app to control your mouse and keyboard.

https://github.com/dessant/buster-client/wiki/Installing-the-client-app#macos

**Comment by adamwinn** [comment]
It's setup correctly and was working fine, until it stopped one day. I dont know when as I dont solve many captchas.  It was working fine before but now it's not.

**Comment by dessant** [comment]
Then it's most likely an upstream issue with robotgo, but unfortunately I cannot debug it due to lack of hardware. Alternatively we could try to downgrade robotgo, but that would reintroduce fixed bugs for other users.

If anyone would like to donate a Mac mini for the project or something similar to support Buster properly on macOS, please contact me. If you have time to contribute, a pull request with a fix would also be welcome.

**Comment by adamwinn** [comment]
Couldn't you install a VM with macOS?

**Comment by dsedivec** [comment]
@dessant Sorry for jacking this thread, I should have realize it was about Linux.  I'll try reinstalling the client, and if I can't find anything useful I'll just disable Buster.  Totally understand about not having a Mac, thank you for this project!  It's really been great.

@adamwinn:

> Couldn't you install a VM with macOS?

That is probably both a violation of Apple's EULA and copyright law.  Not to mention a lot of work.

**Comment by dessant** [comment]
That would break their TOS, and I also can't get a clean copy of macOS to install in a VM, without putting my trust in random people on the internet to download a VM image.

**Comment by dsedivec** [comment]
Since I've already sinned, maybe I can partially redeem myself by saying: I got Buster working again.

Let me detail the things I did, in the order I did them, but **I suspect this first set of steps were unnecessary**:

1. I deleted `~/.local/opt/buster/buster-client` and `~/Library/Application Support/Mozilla/NativeMessagingHosts/org.buster.client.json`.
2. I restarted Firefox.
3. I told Buster to download the client and then I followed the [installation instructions](https://github.com/dessant/buster-client/wiki/Installing-the-client-app#macos), selecting all the defaults when the install screen popped up.

At this point Buster still didn't work, so then **I did what I think really mattered**:

1. I went into System Preferences → Security & Privacy → Privacy → Accessibility and deleted `buster-client`, as in I pushed the little "-" (minus) sign at the bottom to remove it from the list entirely.  (This is in contrast to unchecking it.)
2. Then I clicked the "+" (plus) and re-added `~/.local/opt/buster/buster-client` to Accessibility.
3. I restarted Firefox.

After this, I was able to use Buster at [this reCAPTCHA test](https://patrickhlauke.github.io/recaptcha/).

"But why did you need to re-add `buster-captcha`?  It was already in the list!"  This is an excellent question.  ¯\\_(ツ)_/¯ #JustAppleThings file a radar lol

@dessant, thanks again!  I am *so tired* of arguing with Google's AI over what is or is not a crosswalk.

**Comment by adamwinn** [comment]
There still should be a try and catch to make sure `o` is not undefined and give some sort of error message, rather than it just puking.  `TypeError: o is undefined`

**Comment by dessant** [comment]
@dsedivec, thanks for the detailed instructions, you hereby fully redeemed yourself!

People were telling me that the browser could also be added to the accessibility list, maybe that would sidestep this issue after client app updates.

@adamwinn, I think robotgo silently fails to move the mouse and there is no error to catch, though it would be great if someone would check it, and if needed, create a pull request.

**Comment by adamwinn** [comment]
>People were telling me that the browser could also be added to the accessibility list, maybe that would sidestep this issue after client app updates.

I already had firefox in the accessibility list

**Comment by dessant** [comment]
@adamwinn, though you're right that on the JS side we could check if any mouse events have been dispatched since we asked the client app to move the mouse, and show a warning with instructions if there were no mouse events.

**Comment by adamwinn** [comment]
So if doing a manual uninstall fixed this, then maybe a proper, full uninstall script could be created which would remove `~/.local/opt/buster/buster-client` and others, and then during the install process afterwards would then prompt to have `buster-client` added to the accessibility list, just like as if it were a brand new install

Or at the least make a wiki entry saying for macOS to remove and add `buster-client` to the accessibility list

**Comment by rootedetc** [comment]
I've followed all the steps by @dsedivec however I haven't been able to use Buster.
I've removed both Buster and Buster-client about 5 times, and also flushed it from the Security & privacy option in Settings (MacOS)

I've tried using WIT speech managed api and IBM watson, same problem every time.

Here is what happens: upon pressing the Buster button, pointer gets dragged all the way to the top left corner, Buster hangs for 5 seconds before spitting out _TypeError: o is undefined_ in console.

I've been able to use Buster on MacOS right up until sometime in early February.
